src/main/kotlin/com/coder/gateway/CoderRemoteConnectionHandle.kt

```diff
@@ -0,0 +1,204 @@
+                        if (link != null) browserLink(link.first, link.second)
+                        inputTextField = textField()
+                            .applyToComponent {
+                                text = default ?: ""
+                                minimumSize = Dimension(520, -1)
+                            }.component
+                    }.layout(RowLayout.PARENT_GRID)
+                    row {
+                        cell() // To align with the text box.
+                        cell(
+                            ComponentPanelBuilder.createCommentComponent(comment, false, -1, true)
+                                .applyIf(isError) {
+                                    apply {
+                                        foreground = UIUtil.getErrorForeground()
+                                    }
+                                }
+                        )
+                    }.layout(RowLayout.PARENT_GRID)
+                }
+                AppIcon.getInstance().requestAttention(null, true)
+                if (!dialog(
+                        CoderGatewayBundle.message("gateway.connector.view.login.token.dialog"),
+                        panel = panel,
+                        focusedComponent = inputTextField
+                    ).showAndGet()
+                ) {
+                    return@invokeAndWait
+                }
+                inputFromUser = inputTextField.text
+            }, ModalityState.any())
+            return inputFromUser
+        }
+
+        /**
+         * Open a dialog for providing the token.  Show any existing token so the
+         * user can validate it if a previous connection failed.  If we are not
+         * retrying and the user has not checked the existing token box then open a
+         * browser to the auth page.  If the user has checked the existing token box
+         * then populate the dialog with the token on disk (this will overwrite any
+         * other existing token) unless this is a retry to avoid clobbering the
+         * token that just failed.  Return the token submitted by the user.
+         */
+        @JvmStatic
+        fun askToken(
+            url: URL,
+            token: Pair<String, TokenSource>?,
+            isRetry: Boolean,
+            useExisting: Boolean,
+        ): Pair<String, TokenSource>? {
+            var (existingToken, tokenSource) = token ?: Pair("", TokenSource.USER)
+            val getTokenUrl = url.withPath("/login?redirect=%2Fcli-auth")
+            if (!isRetry && !useExisting) {
+                BrowserUtil.browse(getTokenUrl)
+            } else if (!isRetry && useExisting) {
+                val (u, t) = CoderCLIManager.readConfig()
+                if (url == u?.toURL() && !t.isNullOrBlank() && t != existingToken) {
+                    logger.info("Injecting token for $url from CLI config")
+                    tokenSource = TokenSource.CONFIG
+                    existingToken = t
+                }
```

Comment:
I think the below table describes the behaviour we want:

```
isRetry | useExisting | action
===============================
F       | F           | ask
F       | T           | inject
T       | F           | ask
T       | T           | ask (because it's a retry and we must have failed previously)
```

So I think you can just drop the extra `if` part of the else clause:

```
if (!isRetry && useExisting) {
  // inject
} else {
  // ask
}
```

Reply:
Ahh, so basically skip the ask when we inject?  I originally wanted to ask in that case so they could choose whether to accept the token from their config or enter a new one, but maybe it makes more sense to jump ahead and just use it?

Reply:
TBH that's the behaviour I would expect!